src/tools/Line.ts

```diff
@@ -0,0 +1,159 @@
+    canvasElement.data = [e.evt.x, e.evt.y]
+    canvasElement.id = uuid()
+    canvasElement.endStyle = this.endStyle
+    canvasElement.strokeStyle = this.strokeStyle
+    this.line = new Konva.Line()
+    this.arrow = new Konva.Arrow()
+    this.TBar = new Konva.Line()
+    if (this.endStyle === 'line') {
+      this.line = this.createLineElement(canvasElement)
+      layer.add(this.line)
+    } else if (this.endStyle === 'arrow') {
+      this.arrow = this.createArrowElement(canvasElement)
+      layer.add(this.arrow)
+    } else if (this.endStyle === 'T') {
+      this.line = this.createLineElement(canvasElement)
+      this.TBar = this.createTElement(canvasElement)
+      layer.add(this.TBar)
+      layer.add(this.line)
+    }
+  }
+
+  // eslint-disable-next-line
+  mouseMoveAction = throttle((e: Konva.KonvaPointerEvent, canvasElement: CanvasElement, layer: Konva.Layer, socket: WebSocket): void => {
+    this.TBar.points([canvasElement.data[0], canvasElement.data[1], e.evt.x, e.evt.y])
+    if (this.endStyle === 'line') {
+      this.line.points([canvasElement.data[0], canvasElement.data[1], e.evt.x, e.evt.y])
+    } else if (this.endStyle === 'arrow') {
+      this.arrow.points([canvasElement.data[0], canvasElement.data[1], e.evt.x, e.evt.y])
+    } else if (this.endStyle === 'T') {
+      this.TBar.points(this.calcTBar(canvasElement.data[0], canvasElement.data[1], e.evt.x, e.evt.y))
+      this.line.points([canvasElement.data[0], canvasElement.data[1], e.evt.x, e.evt.y])
+    }
+    layer.batchDraw()
+  }, 5)
+
+  mouseUpAction = (e: Konva.KonvaPointerEvent, canvasElement: CanvasElement, _layer: Konva.Layer, socket: WebSocket): void => {
+    canvasElement.data = canvasElement.data.concat([e.evt.x, e.evt.y])
+    this.sendToWebsockets(socket, canvasElement)
+  }
+
+  renderCanvas = (canvasElement: CanvasElement, layer: Konva.Layer): void => {
+    this.strokeStyle = canvasElement.strokeStyle || 0
+    if (canvasElement.endStyle === 'line') {
+      layer.add(this.createLineElement(canvasElement, canvasElement.tool.colour, canvasElement.tool.size))
+    } else if (canvasElement.endStyle === 'arrow') {
+      layer.add(this.createArrowElement(canvasElement, canvasElement.tool.colour, canvasElement.tool.size))
+    } else if (canvasElement.endStyle === 'T') {
+      layer.add(this.createLineElement(canvasElement, canvasElement.tool.colour, canvasElement.tool.size))
+      layer.add(this.createTElement(canvasElement, canvasElement.tool.colour, canvasElement.tool.size).points(
+        this.calcTBar(canvasElement.data[0], canvasElement.data[1], canvasElement.data[2], canvasElement.data[3]))
+      )
+    } else {
+      layer.add(this.createLineElement(canvasElement, canvasElement.tool.colour, canvasElement.tool.size))
+    }
+    layer.add(this.createLineElement(canvasElement, canvasElement.tool.colour, canvasElement.tool.size))
```

Comment:
You are adding a line element twice, is this needed?

Reply:
whoops